metatron/src/main/java/uk/gov/ida/eidas/metatron/domain/MetadataResolverService.java

```diff
@@ -0,0 +1,78 @@
+package uk.gov.ida.eidas.metatron.domain;
+
+import org.glassfish.jersey.client.JerseyClientBuilder;
+import org.opensaml.saml.metadata.resolver.MetadataResolver;
+import org.opensaml.saml.metadata.resolver.filter.MetadataFilter;
+import org.opensaml.saml.metadata.resolver.impl.AbstractReloadingMetadataResolver;
+import uk.gov.ida.eidas.metatron.health.MetadataHealth;
+import uk.gov.ida.saml.metadata.ExpiredCertificateMetadataFilter;
+import uk.gov.ida.saml.metadata.PKIXSignatureValidationFilterProvider;
+import uk.gov.ida.saml.metadata.factories.MetadataResolverFactory;
+
+import javax.ws.rs.client.Client;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class MetadataResolverService {
+
+    private static long MIN_REFRESH_DELAY_MS = 60_000;
+    private static long MAX_REFRESH_DELAY_MS = 600_000;
+
+    private Map<String, MetadataResolver> metadataResolverMap;
+    private ExpiredCertificateMetadataFilter expiredCertificateMetadataFilter;
+    private MetadataResolverFactory metadataResolverFactory;
+
+    public MetadataResolverService(EidasConfig countriesConfig) {
+        this.metadataResolverFactory = new MetadataResolverFactory();
+        this.metadataResolverMap = new HashMap<>();
+        this.expiredCertificateMetadataFilter = new ExpiredCertificateMetadataFilter();
+
+        countriesConfig.getCountries().forEach(country -> {
+                    metadataResolverMap.put(country.getEntityId(), getMetadataResolver(country));
+                }
+        );
+    }
+
+    public String getLastRefresh(String entityId) {
+        return ((AbstractReloadingMetadataResolver) this.metadataResolverMap.get(entityId)).getLastRefresh().toString();
+    }
+
+    public List<MetadataHealth> getHealth() {
+        List<MetadataHealth> metadataHealth = new ArrayList<>();
+        this.metadataResolverMap.keySet().stream().forEach(k -> metadataHealth.add(new MetadataHealth(k, (AbstractReloadingMetadataResolver) metadataResolverMap.get(k))));
+        return metadataHealth;
+    }
+
+    public MetadataResolver getMetadataResolver(String entityId) {
+        return metadataResolverMap.get(entityId);
+    }
+
+    private Client getClient(EidasCountryConfig country) {
+        if (Objects.isNull(country.getTlsTruststore())) {
```

Comment:
Maybe `getTlsTruststore()` should be returning an optional?

Reply:
I think that makes the relatively simple json to pojo conversion a little more complex at the EidasCountryConfig level and we'd still end up having to do some kind of check on whether the Optional was empty or not.  As this method is only used once (per client created) when the ResolverService is spun up I think it's okay to have it as it is.  If it were used as part of a public API I think there's more of a case for it but in this situation it's okay.

Reply:
I changed my mind about this and went with the Optional getter.  It _is_ better to signal that there may not be a TLS Truststore and actually, as long as it's just the getter that returns the optional, my concerns about the jackson side being more complicated are nonsense.